c/extras/py_ftrl.cc

```diff
@@ -0,0 +1,413 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+//------------------------------------------------------------------------------
+#include "extras/py_ftrl.h"
+#include "frame/py_frame.h"
+#include "python/float.h"
+#include "python/int.h"
+
+namespace py {
+
+PKArgs Ftrl::Type::args___init__(0, 0, 9, false, true,
+                                 {"a", "b", "l1", "l2", "d", "n_epochs",
+                                 "inter", "hash_type", "seed"},
+                                 "__init__", nullptr);
+
+void Ftrl::m__init__(PKArgs& args) {
+  FtrlModelParams fmp = FtrlModel::fmp_default;
+
+  if (!(args[0].is_undefined() || args[0].is_none())) {
+    fmp.a = args[0].to_double();
+  }
+
+  if (!(args[1].is_undefined() || args[1].is_none())) {
+    fmp.b = args[1].to_double();
+  }
+
+  if (!(args[2].is_undefined() || args[2].is_none())) {
+    fmp.l1 = args[2].to_double();
+  }
+
+  if (!(args[3].is_undefined() || args[3].is_none())) {
+    fmp.l2 = args[3].to_double();
+  }
+
+  if (!(args[4].is_undefined() || args[4].is_none())) {
+    fmp.d = static_cast<uint64_t>(args[4].to_size_t());
+  }
+
+  if (!(args[5].is_undefined() || args[5].is_none())) {
+    fmp.n_epochs = args[5].to_size_t();
+  }
+
+  if (!(args[6].is_undefined() || args[6].is_none())) {
+    fmp.inter = args[6].to_bool_strict();
+  }
+
+  if (!(args[7].is_undefined() || args[7].is_none())) {
+    fmp.hash_type = static_cast<unsigned int>(args[7].to_size_t());
+  }
+
+  if (!(args[8].is_undefined() || args[8].is_none())) {
+    fmp.seed = static_cast<unsigned int>(args[8].to_size_t());
+  }
+
+  if (args.num_varkwd_args() > 0) {
+    throw ValueError() << "Unexpected keyword argument";
+  }
+
+  fm = new FtrlModel(fmp);
+}
+
+
+void Ftrl::m__dealloc__() {
+  delete fm;
+}
+
+
+const char* Ftrl::Type::classname() {
+  return "datatable.core.Ftrl";
+}
+
+
+const char* Ftrl::Type::classdoc() {
+  return R"(Follow the Regularized Leader (FTRL) model with hashing trick.
+    
+See this reference for more details:
+https://www.eecs.tufts.edu/~dsculley/papers/ad-click-prediction.pdf
+
+Parameters
+----------
+a : float
+    `alpha` in per-coordinate learning rate formula.
+b : float
+    `beta` in per-coordinate learning rate formula.
+l1 : float
+    L1 regularization parameter.
+l2 : float
+    L2 regularization parameter.
+d : int
+    Number of bins to be used after the hashing trick. 
+n_epochs : int
+    Number of epochs to train for.
+inter : boolean
+    If feature interactions to be used or not.
+hash_type : int
+    Hashing method to use for strings:
+    `0` - std::hash;
+    `1` - Murmur2;
+    `2` - Murmur3.
+seed: unsigned int
+    Seed to be used for Murmur hash functions.
```

Comment:
Ok for boolean. But how do you indicate only positive integers are allowed here if you just say `int`?

Reply:
You can say so in text. For example, `d` and `n_epochs` are positive too, but you don't specify them as "unsigned". Besides, nothing bad will happen if you allow negative seeds...

Reply:
Yes, need to specify this for `d` and `n_epochs` too. Sure, negative seed won't break anything because at some point it will be converted to the positive one, but negative `d` and `n_epochs`, for example, won't work. If your convention is just to say it in text, I will change it to be like that. Thanks.